Implement spaces list

JammingBen opened the pull request:
## Description
Adds an entry point to access all available spaces for the user.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes https://github.com/owncloud/web/issues/6104

## Screenshots (if appropriate):

![image](https://user-images.githubusercontent.com/50302941/147481658-c3c9a65f-1382-40c5-8927-411be877766d.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [x] Implement suitable software architecture
- [ ] Design (?)
- [x] Tests
- [x] Changelog item

JammingBen commented:
This is ready for a first review cycle. A few annotations:

* We're not sure about the routing. The way we implemented this causes issues with the active flag. Assuming the spaces live under the `/spaces/projects`, this has to be fixed.
![image](https://user-images.githubusercontent.com/50302941/147934507-047a15ba-26cd-4cd6-a829-01ac1cb3404a.png)
* The missing spaces icon in the left sidebar requires the new ODS release.
* URL for `Learn more about spaces` is still missing.

kulmann commented:
@dragotin how's the `learn more about spaces` link planned? Is there already some info page on owncloud.com? Or can we at least define a URL for it (without the page actually existing) which can be placed in the code?

kulmann commented on packages/web-app-files/src/index.js:
```diff
@@ -72,6 +73,13 @@ const navItems = [
       path: `/${appInfo.id}/shares/via-link`
     }
   },
+  {
```
Can't be a static nav item. You need to check the capabilities if the spaces feature is enabled. At the moment this nav item will be visible in an oc10 deployment as well.

JammingBen replied:
There is already a `enabled` prop available, so we used this one (see https://github.com/owncloud/web/pull/6199/files#diff-596b5f5cdc5da8df260bef258a7439ac877ae23495400c227bb167954fdabcb1R82). Technically the route is still accessible, but not showing in the side bar. Is this sufficient, or should be prevent access to it as well?

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+      <list-loader v-if="loadSpacesTask.isRunning" />
+      <template v-else>
+        <div
+          v-if="!!spaces.length"
```
```suggestion
          v-if="spaces.length"
```
Strict check has no benefit here IMO.

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+            <span class="spaces-list-card oc-border uk-card uk-card-default">
+              <span class="uk-card-media-top oc-border-b">
+                <img
+                  :src="space.image ? space.image : defaultImg"
```
Can be simplified to
```suggestion
                  :src="space.image || defaultImg"
```

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+        <div v-else>
+          <no-content-message id="files-spaces-empty" class="files-empty" icon="space">
+            <template #message>
+              <span v-translate>You don't have any spaces currently</span>
```
Other views have a combination of big icon  + message for the no content message. Could you do the same here as well?
Also, text could be `You don't have access to any spaces`. The point in time (`at the moment` or `currently`) is not needed in the message. Could be streamlined / simplified in other views as well but doesn't need to happen in this PR.

JammingBen replied:
Adjusted the text 👍  The icon should already be displayed due to `icon="space"` as soon as the ODS gets updated.

kulmann replied:
ah, perfect, thanks!

kulmann reviewed (changes requested):
The static nav item is an issue. We need to have an enabled spaces capability. IMO we should add an optional `isEnabled` callback to nav items, similar to the file action mixins. Could you look into that? Needs the store in the param object so that you can check the capabilities.

kulmann commented on packages/web-app-files/src/index.js:
```diff
@@ -72,6 +73,16 @@ const navItems = [
+      path: `/${appInfo.id}/spaces/projects`
+    },
+    enabled(capabilities) {
+      return capabilities.core.status.productname !== 'ownCloud'
```
Please don't introduce checks on the product name. ;-) we need to introduce a capability for spaces in ocis. Could you have a look at e.g. the archiver or app_provider capabilities in ocis and do a similar thing for spaces in a PR to the ocis repo? I'd like to have it versioned, same as with archivers and app_providers.

kulmann replied:
should exist already... maybe you can just use it https://github.com/cs3org/reva/pull/2015

ownclouders commented:
Results for <strong>oC10SharingExternalRoot</strong> https://drone.owncloud.com/owncloud/web/21530/41/1
The following scenarios passed on retry:
- webUISharingExternalToRoot/federationSharing.feature:378

ownclouders commented:
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/21540/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

ownclouders commented:
Results for <strong>oCISSharingInternal3</strong> https://drone.owncloud.com/owncloud/web/21543/59/1
The following scenarios passed on retry:
- webUISharingInternalGroupsSharingIndicator/shareWithGroups.feature:136

ownclouders commented:
Results for <strong>oC10SharingPublic2</strong> https://drone.owncloud.com/owncloud/web/21543/38/1
The following scenarios passed on retry:
- webUISharingPublicDifferentRoles/shareByPublicLinkDifferentRoles.feature:286

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+                <img
+                  :src="space.image || defaultImg"
+                  :class="{ 'spaces-list-default-img': !space.image }"
+                  :alt="$gettext('Space image')"
```
Image is only presentational. Should have an empty `alt` tag (`alt=""`) for a11y.

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+            </span>
+          </a>
+        </div>
+        <div v-else>
```
The `no-content-message` component doesn't need a wrapping `div`.
Also, the other views have the following structure:
```
list-load v-if...
template v-else
   no-content-message v-if=empty
   whatever-you-want-to-display
```
Would be good if you keep the same structure, i.e. move the `no-content-message` above the actual spaces overview please.

JammingBen replied:
But then the loading spinner would overlay the whole page, not just `Your spaces`. Wouldn't it be nice to always show the header section with `Access all project related spaces...`? Same goes for the no-content-message.

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+          </a>
+        </div>
+        <div v-else>
+          <no-content-message id="files-spaces-empty" class="files-empty" icon="space">
```
If you rebase to current master, you can use `layout-grid` as an icon now.

kulmann replied:
Also, would be good to have the no content message vertically centered, like in the other views.

kulmann commented on packages/web-app-files/src/index.js:
```diff
@@ -72,6 +73,16 @@ const navItems = [
   },
+  {
+    name: $gettext('Spaces'),
+    iconMaterial: 'space',
```
Please use `layout-grid` after rebase.

kulmann reviewed (changes requested):
Some nitpicks. ;-)
Please rebase to current master. Icons/ODS update has been merged to web master.

In `index.js` of the files app you can append `personal/` to the route of the `All files` nav item. That will fix the duplicate highlighting of active nav items in the sidebar.

kulmann commented:
> * Should we hide the view options?
Makes sense, yes.

> * Is UI Kit being removed in the near future? Then we have change the classes being used.
Yes, planned to be removed this week. ;-) basically anything that is `uk-` will be `oc-` then AFAIK.  @pascalwengerter can say more about this.

kulmann commented on packages/web-app-files/src/index.js:
```diff
@@ -72,6 +73,16 @@ const navItems = [
+      path: `/${appInfo.id}/spaces/projects`
+    },
+    enabled(capabilities) {
+      return capabilities.spaces && capabilities.spaces.enabled === true
```
Can be shortened to `capabilities.spaces?.enabled`

JammingBen replied:
Unfortunately not: `SyntaxError: Unexpected token (85:54) in /Users/jannik/projects/owncloud/web/packages/web-app-files/src/index.js`

kulmann replied:
Ok, then keep it like is 😞

pascalwengerter commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  &-default-img {
+    padding: 40px;
```
Could be simplified by giving the `oc-my-m` and `oc-mx-m` helper classes on the img directly

pascalwengerter commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,107 @@
+<template>
+  <div class="oc-p-s">
+    <h2 class="" v-text="$gettext('Spaces')"></h2>
```
Maybe remove the empty classes if not needed? Tag could also be self-closing (just nitpicking here)

JammingBen commented:
Addressed your comments 👍  Two questions left:

* See https://github.com/owncloud/web/pull/6199#discussion_r778655011 -> Should we keep the spaces "header" regardless of the no-content-message? If so, it's hard to 100% center the message vertically, because `height: 100%` does not work. We added `50%` for now, it's better, but not exactly centered.
* We removed the view options. Now the app bar is empty but still takes away some space. Okay, or hide as well? I think the spaces won't use the existing app bar at all, but rather their "own" implementation.
![image](https://user-images.githubusercontent.com/50302941/148199154-3957b6a6-1101-459b-af29-385fae42f8c9.png)

kulmann commented:
> * See [Implement spaces list #6199 (comment)](https://github.com/owncloud/web/pull/6199#discussion_r778655011) -> Should we keep the spaces "header" regardless of the no-content-message? If so, it's hard to 100% center the message vertically, because `height: 100%` does not work. We added `50%` for now, it's better, but not exactly centered.

Awesome, looks better now. 👍 

> * We removed the view options. Now the app bar is empty but still takes away some space. Okay, or hide as well? I think the spaces won't use the existing app bar at all, but rather their "own" implementation.

I hope that we can use the same AppBar for spaces as for the other views, but it will probably need some restructuring. ;-) For the overview of the spaces I think we can hide it completely.

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,109 @@
+                />
+              </span>
+              <span class="uk-card-body">
+                <h3 class="uk-card-title">{{ space.name }}</h3>
```
The h3 must be a h4 (it's a child of the `Your spaces` h3).

kulmann commented on packages/web-app-files/src/spaces/views/Spaces.vue:
```diff
@@ -0,0 +1,109 @@
+      <a href="#" v-text="$gettext('Learn more about spaces.')" />
+      <h3 v-text="$gettext('Your spaces')" />
+      <hr class="oc-mb-s" />
+    </div>
```
The div should wrap the entire content of the spaces list. I.e. span, a, h3, hr must be on the same level as the list-loader and the template.

fschade commented on packages/web-app-files/src/router/spaces.ts:
```diff
@@ -15,15 +15,42 @@ export const createLocationSpaces = (name: shareTypes, location = {}): Location
     path: '/spaces',
     redirect: (to) => createLocationSpaces('files-spaces-personal-home', to)
   },
+  {
```
is this needed since we have `/spaces/:namespace`?

JammingBen replied:
We didn't manage to make it work this way. Maybe you could have a look?

fschade replied:
i tested it and have the same problem due BASE_PATH. lets keep it for now as it is

fschade commented on packages/web-app-files/src/spaces/sdk/index.ts:
```diff
@@ -0,0 +1,30 @@
+
+  return {
+    drives: {
+      listMyDrives: (): AxiosPromise<CollectionOfDrives> => meDrivesApi.listMyDrives()
```
can be `listMyDrives: meDrivesApi.listMyDrives`

JammingBen replied:
This is not working for me, spaces list will be empty then :(

fschade replied:
i can confirm, it errors due the BASE_PATH. Lets keep it as it is

fschade commented on packages/web-app-files/tests/unit/spaces/Spaces.spec.js:
```diff
@@ -0,0 +1,75 @@
+
+function getMountedWrapper() {
+  return mount(Spaces, {
+    localVue,
```
localVue is only used in getMountedWrapper, can we move the initialization into the function?

JammingBen replied:
Are you sure this works? The tests won't run for me then. `localVue` is not being initialized in the spec file, we just import `import { localVue } from '../views/views.setup'`

fschade replied:
I missed that, thanks

fschade commented on packages/web-app-files/tests/unit/spaces/Spaces.spec.js:
```diff
@@ -0,0 +1,75 @@
+    router: new VueRouter(),
+    store: createStore(Vuex.Store, {
+      getters: {
+        configuration: jest.fn(() => ({
```
no jest.fn needed

fschade reviewed (commented):
found some small things.
I would love to see some e2e tests as part of this pr.

fschade commented:
needs to be squashed before merging, @JammingBen can you have a look?

ownclouders commented:
Results for <strong>oCISSharingInternal3</strong> https://drone.owncloud.com/owncloud/web/21850/59/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

ownclouders commented:
Results for <strong>oC10IntegrationApp1</strong> https://drone.owncloud.com/owncloud/web/21862/68/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

pascalwengerter commented on packages/web-app-files/tests/unit/views/spaces/Projects.spec.js:
```diff
@@ -0,0 +1,62 @@
+    const wrapper = getMountedWrapper()
+    await wrapper.vm.loadSpacesTask.last
+
+    expect(wrapper.vm.spaces.length).toEqual(1)
```
You could add a snapshot here (`expect(wrapper).toMatchSnapshot()`) to make sure the DOM stays the way we expect it?

kulmann reviewed (approved):
Awesome! 💪

kulmann commented:
One thing that I found is, that the spaces are neither sorted by created time, nor by name. Not an issue for this PR, but please open a ticket about default sorting or simply sort by name for now.

refs commented on packages/web-client/package.json:
```diff
@@ -0,0 +1,11 @@
+  "license": "AGPL-3.0",
+  "main": "src/index.ts",
+  "scripts": {
+    "generate-openapi": "rm -rf src/generated && docker run --rm -v \"${PWD}/src:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v0.0.yaml  -g typescript-axios -o /local/generated"
```
why not using [this](https://github.com/owncloud/libre-graph-api-typescript-axios)?

see https://github.com/owncloud/libre-graph-api/blob/main/.drone.star#L17-L21

JammingBen commented:
> One thing that I found is, that the spaces are neither sorted by created time, nor by name. Not an issue for this PR, but please open a ticket about default sorting or simply sort by name for now.

I've created https://github.com/owncloud/web/issues/6253, since we *really* want to merge this now as the pipeline is currently green for once :D

JanAckermann commented:
> One thing that I found is, that the spaces are neither sorted by created time, nor by name. Not an issue for this PR, but please open a ticket about default sorting or simply sort by name for now.

Sorting should be a task for the API, this feature is yet not there, so we just wait until it is done and fix this in the next PR :innocent:

kulmann commented:
> Sorting should be a task for the API, this feature is yet not there, so we just wait until it is done and fix this in the next PR :innocent: 

Yes, of course something for the API. I meant it as a mitigation for the status quo. Which is a oneliner. ;-) if sorting in the backend happens soon then it's fine.
